File: internal/session/session.go
// Package session defines argus's transport- and tool-agnostic model for every AI
// coding tool session it discovers or manages.
package session

// Status is the lifecycle/activity state of a session, driven by tool hooks where
// available and a tmux fallback heuristic otherwise.
type Status string

const (
	// StatusDiscovered: found (e.g. a tmux pane) but no hook event yet.
	StatusDiscovered Status = "discovered"
	// StatusWorking: actively processing (between prompt submit and stop, or a tool call).
	StatusWorking Status = "working"
	// StatusAwaitingInput: blocked on the user (e.g. a permission prompt).
	StatusAwaitingInput Status = "awaiting_input"
	// StatusIdle: finished responding, waiting for the next prompt.
	StatusIdle Status = "idle"
	// StatusDead: the underlying process/pane is gone.
	StatusDead Status = "dead"
)

// Label is the display word clients show next to the glyph. The server owns this
// string so wording stays consistent across clients.
func (s Status) Label() string {
	switch s {
	case StatusWorking:
		return "working"
	case StatusAwaitingInput:
		return "awaiting"
	case StatusIdle:
		return "idle"
	case StatusDiscovered:
		return "discovered"
	case StatusDead:
		return "dead"
	default:
		return string(s)
	}
}

// Source records how argus learned about a session.
type Source string

const (
	// SourceDiscovered: found by scanning tmux.
	SourceDiscovered Source = "discovered"
	// SourceSpawned: created by argus on its private tmux socket.
	SourceSpawned Source = "spawned"
	// SourceHooked: first seen via a tool hook event.
	SourceHooked Source = "hooked"
)

// Frontend classifies where a session's UI lives, which determines whether argus
// can drive its terminal (only tmux panes are controllable).
type Frontend string

const (
	// FrontendTmux: has a tmux pane → fully controllable.
	FrontendTmux Frontend = "tmux"
	// FrontendVSCode: paneless, started from the VSCode extension (entrypoint "claude-vscode").
	FrontendVSCode Frontend = "vscode"
	// FrontendExternal: paneless, some other non-tmux terminal.
	FrontendExternal Frontend = "external"
)

// TmuxServer identifies which tmux server a pane lives on.
type TmuxServer string

const (
	// TmuxServerDefault is the user's normal tmux server.
	TmuxServerDefault TmuxServer = "default"
	// TmuxServerArgus is the private "tmux -L argus" socket argus spawns into.
	TmuxServerArgus TmuxServer = "argus"
)

// TmuxLocation pins a session to a concrete tmux pane. PaneID (e.g. "%3") is the
// stable, never-reused identifier and is the primary correlation key.
type TmuxLocation struct {
	Server      TmuxServer `json:"server"`
	PaneID      string     `json:"pane_id"`
	SessionName string     `json:"session_name"`
	WindowIndex int        `json:"window_index"`
	CurrentPath string     `json:"current_path"`
}

// InteractionKind classifies what a session is waiting on the user for.
type InteractionKind string

const (
	// InteractionPermission: Claude wants approval to use a tool (1/2/3 prompt).
	InteractionPermission InteractionKind = "permission"
	// InteractionQuestion: an AskUserQuestion multiple-choice prompt.
	InteractionQuestion InteractionKind = "question"
	// InteractionPlan: an ExitPlanMode plan-approval prompt.
	InteractionPlan InteractionKind = "plan"
	// InteractionIdle: Claude finished and is waiting for the next message, or a
	// generic notification needing attention.
	InteractionIdle InteractionKind = "idle"
)

// QuestionSpec is one question within an AskUserQuestion interaction (a call may
// carry 1-4, each independently answered).
type QuestionSpec struct {
	Header      string `json:"header,omitempty"`       // short category label (chip)
	Question    string `json:"question,omitempty"`     // the question prompt
	MultiSelect bool   `json:"multi_select,omitempty"` // allows multiple selections
	// Options and the positionally-aligned per-option help/preview text.
	Options            []string `json:"options,omitempty"`
	OptionDescriptions []string `json:"option_descriptions,omitempty"`
	OptionPreviews     []string `json:"option_previews,omitempty"`
}

// DecisionOption is a server-built choice for a permission/plan decision. The client
// renders Label and echoes Value back; the node maps Value to the hook decision.
type DecisionOption struct {
	Label  string `json:"label"`
	Value  string `json:"value"`
	Reject bool   `json:"reject,omitempty"` // denies; surfaces a free-text reason field
	// Placeholder prompts the reject free-text field (e.g. "Tell Claude what to change").
	Placeholder string `json:"placeholder,omitempty"`
}

// Interaction is a pending user request, surfaced so a client can render it natively
// and respond. Set while StatusAwaitingInput, cleared when the session moves on.
type Interaction struct {
	Kind      InteractionKind `json:"kind"`
	Message   string          `json:"message,omitempty"`    // hook notification text
	ToolName  string          `json:"tool_name,omitempty"`  // tool awaiting permission, when known
	ToolInput string          `json:"tool_input,omitempty"` // pending tool input (permission detail)
	// Questions holds the AskUserQuestion question(s) (>=1 when Kind is question).
	Questions []QuestionSpec `json:"questions,omitempty"`
	Plan      string         `json:"plan,omitempty"` // ExitPlanMode plan text
	// Options, when set, are server-built decision choices the client renders verbatim;
	// the chosen Value is sent back unchanged and mapped to allow/deny + permission mode.
	Options []DecisionOption `json:"options,omitempty"`
}

// Summary is a cached transcript digest for list views, computed node-side on hook
// events so clients never parse transcripts. All fields are comparable.
type Summary struct {
	Model        string  `json:"model,omitempty"`         // raw, e.g. "claude-opus-4-8"
	HasContext   bool    `json:"has_context,omitempty"`   // whether ContextPct is meaningful
	ContextPct   float64 `json:"context_pct,omitempty"`   // 0..100, latest turn
	Tokens       int     `json:"tokens,omitempty"`        // latest turn prompt-side token count
	Task         string  `json:"task,omitempty"`          // latest user prompt, first line
	LastActivity string  `json:"last_activity,omitempty"` // RFC3339 of the last chunk
}

// Session is argus's record for a single AI coding tool session.
type Session struct {
	// ID is argus's internal identifier (stable for the session's lifetime).
	ID string `json:"id"`
	// Agent is the coding agent that owns this session (e.g. "claude").
	Agent string `json:"agent"`
	// AgentSessionID is the tool's own session id; empty until a hook fires.
	AgentSessionID string `json:"agent_session_id,omitempty"`
	// Name is the tool's own session name (e.g. Claude's session name), when known.
	Name string `json:"name,omitempty"`

	Tmux TmuxLocation `json:"tmux"`

	// Cwd and TranscriptPath are populated from hook payloads.
	Cwd            string `json:"cwd,omitempty"`
	TranscriptPath string `json:"transcript_path,omitempty"`

	Status Status `json:"status"`
	// StatusLabel is the server-rendered display word for Status (set on emit).
	StatusLabel string `json:"status_label,omitempty"`
	Source      Source `json:"source"`

	// Frontend classifies the session's UI host (tmux/vscode/external).
	Frontend Frontend `json:"frontend,omitempty"`

	// Repo is the basename of the session directory's git repository, when known
	// (path-derived, not from the transcript).
	Repo string `json:"repo,omitempty"`

	// Summary is the cached transcript digest for list views (nil until computed).
	Summary *Summary `json:"summary,omitempty"`

	// Interaction is the pending user request (non-nil only while awaiting input).
	Interaction *Interaction `json:"interaction,omitempty"`

	// Origin fields — populated only when surfaced through a gateway.
	NodeID    string `json:"node_id,omitempty"`    // originating node (also composite ID prefix)
	NodeLabel string `json:"node_label,omitempty"` // human-friendly name (e.g. hostname)
	// Offline marks a session whose node is disconnected: kept visible (greyed) for
	// a grace period before removal.
	Offline bool `json:"offline,omitempty"`
}

// Controllable reports whether argus can drive the session's terminal. Only
// tmux-pane sessions are controllable.
func (s Session) Controllable() bool { return s.Tmux.PaneID != "" }

File: app/lib/models/session.dart
import 'enums.dart';

List<String> _strList(Object? v) =>
    v == null ? const [] : (v as List).map((e) => e as String).toList();

class TmuxLocation {
  final TmuxServerKind server;
  final String paneId;
  final String sessionName;
  final int windowIndex;
  final String currentPath;

  const TmuxLocation({
    required this.server,
    required this.paneId,
    required this.sessionName,
    required this.windowIndex,
    required this.currentPath,
  });

  factory TmuxLocation.fromJson(Map<String, dynamic> j) => TmuxLocation(
        server: tmuxServerFromWire(j['server'] as String?),
        paneId: j['pane_id'] as String? ?? '',
        sessionName: j['session_name'] as String? ?? '',
        windowIndex: (j['window_index'] as num?)?.toInt() ?? 0,
        currentPath: j['current_path'] as String? ?? '',
      );
}

class Summary {
  final String? modelName;
  final String? modelColor;
  final bool hasContext;
  final double contextPct;
  final int tokens;
  final String? task;
  final String? lastActivity;

  const Summary({
    this.modelName,
    this.modelColor,
    this.hasContext = false,
    this.contextPct = 0,
    this.tokens = 0,
    this.task,
    this.lastActivity,
  });

  factory Summary.fromJson(Map<String, dynamic> j) => Summary(
        modelName: j['model_name'] as String?,
        modelColor: j['model_color'] as String?,
        hasContext: j['has_context'] as bool? ?? false,
        contextPct: (j['context_pct'] as num?)?.toDouble() ?? 0,
        tokens: (j['tokens'] as num?)?.toInt() ?? 0,
        task: j['task'] as String?,
        lastActivity: j['last_activity'] as String?,
      );
}

class QuestionSpec {
  final String? header;
  final String? question;
  final bool multiSelect;
  final List<String> options;
  final List<String> optionDescriptions;
  final List<String> optionPreviews;

  const QuestionSpec({
    this.header,
    this.question,
    this.multiSelect = false,
    this.options = const [],
    this.optionDescriptions = const [],
    this.optionPreviews = const [],
  });

  factory QuestionSpec.fromJson(Map<String, dynamic> j) => QuestionSpec(
        header: j['header'] as String?,
        question: j['question'] as String?,
        multiSelect: j['multi_select'] as bool? ?? false,
        options: _strList(j['options']),
        optionDescriptions: _strList(j['option_descriptions']),
        optionPreviews: _strList(j['option_previews']),
      );
}

/// A server-built decision choice (e.g. ExitPlanMode approve variants). The
/// client renders [label] and echoes [value] back; [reject] marks the deny
/// choice, which opens a free-text field prompted by [placeholder].
class DecisionOption {
  final String label;
  final String value;
  final bool reject;
  final String placeholder;

  const DecisionOption({
    required this.label,
    required this.value,
    this.reject = false,
    this.placeholder = '',
  });

  factory DecisionOption.fromJson(Map<String, dynamic> j) => DecisionOption(
        label: j['label'] as String? ?? '',
        value: j['value'] as String? ?? '',
        reject: j['reject'] as bool? ?? false,
        placeholder: j['placeholder'] as String? ?? '',
      );
}

class Interaction {
  final InteractionKind kind;
  final String? message;
  final String? toolName;
  final String? toolInput;
  final List<QuestionSpec> questions;
  final String? plan;
  final List<DecisionOption> options;

  const Interaction({
    required this.kind,
    this.message,
    this.toolName,
    this.toolInput,
    this.questions = const [],
    this.plan,
    this.options = const [],
  });

  factory Interaction.fromJson(Map<String, dynamic> j) => Interaction(
        kind: interactionKindFromWire(j['kind'] as String?),
        message: j['message'] as String?,
        toolName: j['tool_name'] as String?,
        toolInput: j['tool_input'] as String?,
        questions: (j['questions'] as List?)
                ?.map((e) => QuestionSpec.fromJson(e as Map<String, dynamic>))
                .toList() ??
            const [],
        plan: j['plan'] as String?,
        options: (j['options'] as List?)
                ?.map((e) => DecisionOption.fromJson(e as Map<String, dynamic>))
                .toList() ??
            const [],
      );
}

class Session {
  final String id;
  final String agent;
  final String? agentSessionId;
  final String? name;
  final TmuxLocation tmux;
  final String? cwd;
  final String? transcriptPath;
  final SessionStatus status;
  final String statusLabel;
  final SessionSource source;
  final FrontendKind frontend;
  final String? repo;
  final Summary? summary;
  final Interaction? interaction;
  final String? nodeId;
  final String? nodeLabel;
  final bool offline;

  const Session({
    required this.id,
    required this.agent,
    required this.tmux,
    required this.status,
    this.statusLabel = '',
    required this.source,
    this.frontend = FrontendKind.unknown,
    this.agentSessionId,
    this.name,
    this.cwd,
    this.transcriptPath,
    this.repo,
    this.summary,
    this.interaction,
    this.nodeId,
    this.nodeLabel,
    this.offline = false,
  });

  factory Session.fromJson(Map<String, dynamic> j) => Session(
        id: j['id'] as String,
        agent: j['agent'] as String? ?? '',
        agentSessionId: j['agent_session_id'] as String?,
        name: j['name'] as String?,
        tmux: TmuxLocation.fromJson(j['tmux'] as Map<String, dynamic>),
        cwd: j['cwd'] as String?,
        transcriptPath: j['transcript_path'] as String?,
        status: statusFromWire(j['status'] as String?),
        statusLabel: j['status_label'] as String? ?? '',
        source: sourceFromWire(j['source'] as String?),
        frontend: frontendFromWire(j['frontend'] as String?),
        repo: j['repo'] as String?,
        summary: j['summary'] == null
            ? null
            : Summary.fromJson(j['summary'] as Map<String, dynamic>),
        interaction: j['interaction'] == null
            ? null
            : Interaction.fromJson(j['interaction'] as Map<String, dynamic>),
        nodeId: j['node_id'] as String?,
        nodeLabel: j['node_label'] as String?,
        offline: j['offline'] as bool? ?? false,
      );

  /// The label shown wherever a session needs a title: repo, else name, else id.
  String get displayTitle => (repo?.isNotEmpty ?? false)
      ? repo!
      : (name?.isNotEmpty ?? false)
          ? name!
          : id;

  /// Whether argus can drive this session's terminal. Derived from frontend:
  /// only tmux sessions are controllable; vscode/external are decision-only.
  bool get controllable => frontend == FrontendKind.tmux;
}
